Zend/zend_operators.c

```diff
@@ -1840,106 +1840,138 @@ ZEND_API zend_result ZEND_FASTCALL shift_right_function(zval *result, zval *op1,
 ZEND_API zend_result ZEND_FASTCALL concat_function(zval *result, zval *op1, zval *op2) /* {{{ */
 {
     zval *orig_op1 = op1;
-	zval op1_copy, op2_copy;
-
-	ZVAL_UNDEF(&op1_copy);
-	ZVAL_UNDEF(&op2_copy);
+	zend_string *op1_string, *op2_string = NULL;
+	bool op1_is_non_string_or_copy = false;
+	bool op2_is_non_string_or_copy = false;
 
 	do {
-	 	if (UNEXPECTED(Z_TYPE_P(op1) != IS_STRING)) {
+	 	if (EXPECTED(Z_TYPE_P(op1) == IS_STRING)) {
+			op1_string = Z_STR_P(op1);
+		} else {
 	 		if (Z_ISREF_P(op1)) {
 	 			op1 = Z_REFVAL_P(op1);
-	 			if (Z_TYPE_P(op1) == IS_STRING) break;
+	 			if (Z_TYPE_P(op1) == IS_STRING) {
+					op1_string = Z_STR_P(op1);
+					break;
+				}
 	 		}
 			ZEND_TRY_BINARY_OBJECT_OPERATION(ZEND_CONCAT);
-			ZVAL_STR(&op1_copy, zval_get_string_func(op1));
+			op1_string = zval_get_string_func(op1);
 			if (UNEXPECTED(EG(exception))) {
-				zval_ptr_dtor_str(&op1_copy);
+				zend_string_release(op1_string);
 				if (orig_op1 != result) {
 					ZVAL_UNDEF(result);
 				}
 				return FAILURE;
 			}
+			op1_is_non_string_or_copy = true;
 			if (result == op1) {
 				if (UNEXPECTED(op1 == op2)) {
-					op2 = &op1_copy;
+					op2_string = op1_string;
 				}
 			}
-			op1 = &op1_copy;
 		}
 	} while (0);
 	do {
-		if (UNEXPECTED(Z_TYPE_P(op2) != IS_STRING)) {
-	 		if (Z_ISREF_P(op2)) {
-	 			op2 = Z_REFVAL_P(op2);
-	 			if (Z_TYPE_P(op2) == IS_STRING) break;
-	 		}
-			ZEND_TRY_BINARY_OP2_OBJECT_OPERATION(ZEND_CONCAT);
-			ZVAL_STR(&op2_copy, zval_get_string_func(op2));
-			if (UNEXPECTED(EG(exception))) {
-				zval_ptr_dtor_str(&op1_copy);
-				zval_ptr_dtor_str(&op2_copy);
-				if (orig_op1 != result) {
-					ZVAL_UNDEF(result);
+		if (!op2_string) {
+			if (EXPECTED(Z_TYPE_P(op2) == IS_STRING)) {
+				op2_string = Z_STR_P(op2);
+			} else {
+				if (Z_ISREF_P(op2)) {
+					op2 = Z_REFVAL_P(op2);
+					if (Z_TYPE_P(op2) == IS_STRING) {
+						op2_string = Z_STR_P(op2);
+						break;
+					}
 				}
-				return FAILURE;
+				/* hold an additional reference because a userland function could free this */
+				if (!op1_is_non_string_or_copy) {
+					op1_string = zend_string_copy(op1_string);
+					op1_is_non_string_or_copy = true;
+				}
+				ZEND_TRY_BINARY_OP2_OBJECT_OPERATION(ZEND_CONCAT);
+				op2_string = zval_get_string_func(op2);
+				if (UNEXPECTED(EG(exception))) {
+					zend_string_release(op1_string);
```

Comment:
This unconditional release doesn't look right anymore. Previously, this was unconditional because `op1_copy` could only contain the newly allocated string but not existing strings.

Reply:
I believe it is right, lines 1888-1891 will make sure that if `op1_free_string` was false before, it will now be true.
We need to do this because a userland function in line 1893 `op2_string = zval_get_string_func(op2);` can cause the refcount for `op1_string` to decrease.
Therefore, we know that at line 1895 the string is always a copy, and we can unconditionally release it.

Reply:
Ah, right.